src/hammer-vlsi/hammer_vlsi/driver.py

```diff
@@ -806,12 +811,26 @@ def get_hierarchical_settings(self) -> List[Tuple[str, dict]]:
             assert isinstance(list_of_hier_modules, list)
             if len(list_of_hier_modules) == 0:
                 raise ValueError("No hierarchical modules defined manually in manual hierarchical mode")
+            hier_modules = reduce(add_dicts, list_of_hier_modules)
+
             list_of_placement_constraints = self.database.get_setting(
                 "vlsi.inputs.hierarchical.manual_placement_constraints")  # type: List[Dict]
             assert isinstance(list_of_placement_constraints, list)
-            hier_modules = reduce(add_dicts, list_of_hier_modules)
-            combined_raw_placement_dict = reduce(add_dicts, list_of_placement_constraints, {})  # type: Dict[str, dict]
-            hier_placement_constraints = {key: list(map(PlacementConstraint.from_dict, lst))
+            combined_raw_placement_dict = reduce(add_dicts, list_of_placement_constraints, {})  # type: Dict[str, List[Dict[str, Any]]]
+
+            def get_toplevel(d: List[Dict[str, Any]]) -> Optional[Dict[str, Any]]:
+                results = list(filter(lambda x: x["type"] == "toplevel", d))
+                if len(results) == 0:
+                    return None
+                else:
+                    return results[-1]
+
+            toplevels_opt = {k: get_toplevel(v) for k, v in combined_raw_placement_dict.items()}  # type: Dict[str, Optional[Dict[str, Any]]]
+            toplevels = {k: v for k, v in toplevels_opt.items() if v is not None}  # type: Dict[str, Dict[str, Any]]
+            hier_macros = list(map(lambda x: MacroSize(library="", name=x[0], width=x[1]["width"], height=x[1]["height"]), toplevels.items()))
```

Comment:
Is this part of the code unit tested?

Reply:
No

Reply:
Is it hard to add one? Again if it's too much work I can help

Reply:
I'll work on this after my talk, then I think it's G2G.

Reply:
OK pushed some unit tests.